hashes/src/chacha20poly1305/mod.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Combine the ChaCha20 stream cipher with the Poly1305 message authentication code
+//! to form an authenticated encryption with additional data (AEAD) algorithm.
+
+pub mod chacha20;
+pub mod poly1305;
+
+use chacha20::ChaCha20;
+use poly1305::Poly1305;
+
+use core::fmt;
+
+/// Zero array for padding slices.
+const ZEROES: [u8; 16] = [0u8; 16];
+
+/// Errors encrypting and decrypting messages with ChaCha20 and Poly1305 authentication tags.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// Additional data showing up when it is not expected.
+    UnauthenticatedAdditionalData,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::UnauthenticatedAdditionalData => write!(f, "Unauthenticated aad."),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Error::UnauthenticatedAdditionalData => None,
+        }
+    }
+}
+
+/// Encrypt and decrypt content along with a authentication tag.
+#[derive(Debug)]
+pub struct ChaCha20Poly1305 {
+    key: [u8; 32],
+    nonce: [u8; 12],
+}
+
+impl ChaCha20Poly1305 {
+    /// Make a new instance of a ChaCha20Poly1305 AEAD.
+    pub fn new(key: [u8; 32], nonce: [u8; 12]) -> Self {
+        ChaCha20Poly1305 { key, nonce }
+    }
+
+    /// Encrypt content in place and return the poly1305 16-byte authentication tag.
+    ///
+    /// # Arguments
+    ///
+    /// - `content` - Plaintext to be encrypted in place.
+    /// - `aad`     - Optional metadata covered by the authentication tag.
+    ///
+    /// # Returns
+    ///
+    /// The 16-byte authentication tag.
+    pub fn encrypt(self, content: &mut [u8], aad: Option<&[u8]>) -> [u8; 16] {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 1);
+        chacha.apply_keystream(content);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        poly.add(aad);
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - aad_overflow)]);
+        }
+
+        poly.add(content);
+        let text_overflow = content.len() % 16;
+        if text_overflow > 0 {
+            poly.add(&ZEROES[0..(16 - text_overflow)]);
+        }
+
+        let aad_len = aad.len().to_le_bytes();
+        let msg_len = content.len().to_le_bytes();
+        let mut len_buffer = [0u8; 16];
+        len_buffer[..aad_len.len()].copy_from_slice(&aad_len[..]);
+        for i in 0..msg_len.len() {
+            len_buffer[i + aad_len.len()] = msg_len[i]
+        }
+        poly.add(&len_buffer);
+        poly.tag()
+    }
```

Comment:
So if I understand it correctly, the user must serialize upfront to feed the whole message at once. This is not great for performance and some alternative should be provided. The interface will be obviously different. Perhaps something like this:

```rust
pub fn encryptor(&mut self) -> Encryptor<'_> { ... }

pub struct Encryptor<'a> {
...
}

impl<'a> Encryptor {
    pub fn encrypt(&mut self, &mut [u8]);
    pub fn finish(self) -> [u8; 16];
}
```